sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -0,0 +1,3732 @@
+}
+
+// No generic lambdas in C++11, so use this convinence macro.
+// NOTE: to be used in API returning "ParamValue".
+// NOTE: memset is used to clear all bytes in the memory allocated by SYCL RT
+// for value. This is a workaround for the problem when return type of the
+// parameter is incorrect in L0 plugin which can result in bad value. This
+// memset can be removed if it is necessary.
+#define SET_PARAM_VALUE(Value)                                                 \
+  {                                                                            \
+    typedef decltype(Value) T;                                                 \
+    if (ParamValue) {                                                          \
+      memset(ParamValue, 0, ParamValueSize);                                   \
+      *(T *)ParamValue = Value;                                                \
+    }                                                                          \
+    if (ParamValueSizeRet)                                                     \
+      *ParamValueSizeRet = sizeof(T);                                          \
+  }
+#define SET_PARAM_VALUE_STR(Value)                                             \
+  {                                                                            \
+    if (ParamValue)                                                            \
+      memcpy(ParamValue, Value, ParamValueSize);                               \
+    if (ParamValueSizeRet)                                                     \
+      *ParamValueSizeRet = strlen(Value) + 1;                                  \
+  }
+
+#define SET_PARAM_VALUE_VLA(Value, NumValues, RetType)                         \
+  {                                                                            \
+    if (ParamValue) {                                                          \
+      memset(ParamValue, 0, ParamValueSize);                                   \
+      for (uint32_t I = 0; I < NumValues; I++)                                 \
+        ((RetType *)ParamValue)[I] = (RetType)Value[I];                        \
+    }                                                                          \
+    if (ParamValueSizeRet)                                                     \
+      *ParamValueSizeRet = NumValues * sizeof(RetType);                        \
+  }
+
+#ifndef _WIN32
+// Recover from Linux SIGSEGV signal.
+// We can't reliably catch C++ exceptions thrown from signal
+// handler so use setjmp/longjmp.
+//
+#include <setjmp.h>
+#include <signal.h>
+jmp_buf ReturnHere;
+static void piSignalHandler(int SigNum) {
+  // We are somewhere the signall was raised, so go back to
+  // where we started tracking.
+  longjmp(ReturnHere, 0);
+}
+// Only handle segfault now, but can be extended.
+#define __TRY()                                                                \
+  signal(SIGSEGV, &piSignalHandler);                                           \
```

Comment:
AFAIK `signal` will set `piSignalHandler` for `SIGSEGV` for only one time of signal emit.
Suggest using `sigaction`.

Reply:
Thx, fixed. zeInit must be called only once. So I used std::call_once to call zeInit only once and to guarantee that current SIGSEGV signal handling using signal() will be correct (which works only for one time signal emit).